query/call_iterator.go

```diff
@@ -510,6 +512,71 @@ func NewDistinctIterator(input Iterator, opt IteratorOptions) (Iterator, error)
 	}
 }
 
+func newFloatReduceFloatWithPrevElementIterator(input FloatIterator, opt IteratorOptions, createFn func(it FloatPrevIterator) (FloatPointAggregator, FloatPointEmitter)) *floatReduceFloatIterator {
+	prevIt := newPrevElementFloatIterator(input)
+	it := newBufFloatIterator(prevIt)
+	return &floatReduceFloatIterator{
+		input: it,
+		create: func() (FloatPointAggregator, FloatPointEmitter) {
+			return createFn(prevIt)
+		},
+		dims: opt.GetDimensions(),
+		opt:  opt,
+	}
+}
+
+func newIntegerReduceFloatWithPrevElementIterator(input IntegerIterator, opt IteratorOptions, createFn func(it IntegerPrevIterator) (IntegerPointAggregator, FloatPointEmitter)) *integerReduceFloatIterator {
+	prevIt := newPrevElementIntegerIterator(input)
+	it := newBufIntegerIterator(prevIt)
+	return &integerReduceFloatIterator{
+		input: it,
+		create: func() (IntegerPointAggregator, FloatPointEmitter) {
+			return createFn(prevIt)
+		},
+		dims: opt.GetDimensions(),
+		opt:  opt,
+	}
+}
+
+func newUnsignedReduceFloatWithPrevElementIterator(input UnsignedIterator, opt IteratorOptions, createFn func(it UnsignedPrevIterator) (UnsignedPointAggregator, FloatPointEmitter)) *unsignedReduceFloatIterator {
+	prevIt := newPrevElementUnsignedIterator(input)
+	it := newBufUnsignedIterator(prevIt)
+	return &unsignedReduceFloatIterator{
+		input: it,
+		create: func() (UnsignedPointAggregator, FloatPointEmitter) {
+			return createFn(prevIt)
```

Comment:
I think this is wrong, but I'm not aware enough of how time weighted averages work. The one thing I do know is that this seems to be attempting to retrieve a value through state that was intentionally kept separate for this specific iterator.

The `mean()` operator and others relating to it are intentionally separated into their own buckets. They are not allowed to communicate with each other. That's not necessarily correct for this code to use and you'll run into potential bugs if we try to get around that structure.

But, I don't know enough about time weighted average to say which method should be used. So here are the three methods that exist that may be applicable.

1. The method is completely bucketed. The buckets have nothing to do with each other. These are your standard aggregation/selector functions (`mean()`, `max()`, `min()`, `first()`, `last()`, etc).
2. The method has one result per bucket, but the points from a previous bucket may influence the value of the current bucket. The only existing function that does this is `integral()`.
3. The method is not bucketed at all. It consumes a point and produces a point (sometimes). It may produce fewer points than the original stream, but it is mostly 1 point in, 1 point out. The `derivative()`, `difference()`, and `moving_average()` functions are the best example of this. These methods may _operate_ on bucketed data, but they don't care about whether values are in buckets by themselves.

I'm thinking `integral()` is the best example of what you probably need, but I don't know if the result should be bucketed or not.

Reply:
"The method has one result per bucket, but the points from a previous bucket may influence the value of the current bucket." This is valid for time weighted averages. I will try to change the implementation

Reply:
time_weighted_average reimplemented with streaming iterator - similarly to integral